File: src/components/launchers/XboxGames.js
const shell = window.__TAURI__.shell;
const os = window.__TAURI__.os;
const http = window.__TAURI__.http;
const invoke = window.__TAURI__.invoke;

async function getInstalledGames() {
  if ((await os.platform()) === "win32") {
    const { stdout, stderr } = await new shell.Command("cmd", [
      "/C",
      "C:/Windows/System32/WindowsPowerShell/v1.0/powershell.exe",
      "get-appxpackage",
    ]).execute();
    if (stderr) {
      console.log(stderr);
      return [];
    }
    let apps = await parseRawToJSON(stdout).filter(
      (x) =>
        x.IsFramework !== "True" &&
        x.SignatureKind === "Store" &&
        x.PublisherId !== "8wekyb3d8bbwe" &&
        !x.PackageFullName.startsWith("MicrosoftWindows")
    );
    apps = (await Promise.all(verifyGames(apps))).filter(async (x) => {
      return typeof (await x) === "object";
    });
    apps = apps.filter((x) => x?.Banner);
    return (await Promise.all(apps.map((x) => parseGmeObject(x)))).filter(
      (x) => typeof x === "object"
    );
  } else {
    return [];
  }
}

function parseRawToJSON(res) {
  return res
    .split("\r\n\r\n")
    .slice(1)
    .map((x) => {
      return x
        .split("\r\n")
        .filter((y) => !y.startsWith(" "))
        .map((y) =>
          y
            .split(":", 3)
            .map(
              (z, i, arr) =>
                `${
                  arr.length === 3 && i !== 0 ? (i === 2 ? "" : '"') : '"'
                }${z.trim()}${
                  arr.length === 3 && i !== 0 ? (i === 2 ? '"' : "") : '"'
                }`
            )
            .join(":")
        )
        .join(",\n")
        .replaceAll("\\", "/");
    })
    .filter((x) => x.length > 20)
    .map((x) => {
      return JSON.parse(`{${x}}`);
    });
}

async function parseGmeObject(obj) {
  const { Banner, InstallLocation } = obj;
  const Location = InstallLocation;
  const manifest = (
    await invoke("read_file", {filePath: Location + "/AppxManifest.xml"})
  ).split("\r\n");
  const Executable = manifest
    .find((x) => x.trim().startsWith("<Application "))
    .split(/<[/]{0,1}Application[>]{0,1}/)[1]
    .match(/Executable="[a-z0-9\s-]{1,}\.exe"/i);
  if (!Executable) return;

  return {
    Executable: Executable[0].split('"')[1],
    Location,
    DisplayName: obj.Name,
    GameID: obj.PackageFullName,
    LaunchID: `start shell:AppsFolder\\${obj.PackageFamilyName}!${
      manifest
        .find((x) => x.trim().startsWith("<Application "))
        .split(/<[/]{0,1}Application[>]{0,1}/)[1]
        .match(/Id="[a-z]{1,}"/i)[0]
    }`.replace('!Id="App"', "!App"),
    Size: null,
    Banner,
    LauncherName: "XboxGames",
  };
}

function verifyGames(jsons) {
  const arr = [];
  for (let i = 0; i < jsons.length; i++) {
    const json = jsons[i];
    arr.push(
      (async () => {
        json.Name = json.Name.split(".")[1];
        const response = await http
          .fetch(
            `https://www.microsoft.com/en-in/search/shop/games?q=${encodeURIComponent(
              json.Name
            )}&devicetype=pc`,
            { responseType: 2 }
          )
          .catch(() => "");
        if (response === "") return;
        const dom = JSON.parse(
          await invoke("parse", {
            value: response.data,
          })
        );

        const list = dom.children
          .find((x) => x.name === "html")
          .children.find((x) => x.name === "body")
          .children.find((x) =>
            x.attributes
              ? x.attributes["data-grid"] === "container pad-12x stack-2"
              : false
          )
          ?.children.find((x) => x.name === "section")
          .children[0].children.find((x) =>
            x.id ? x.id.includes("productplacementlist") : false
          )?.children;
        if (list.find((x) => x.name === "p")) return false;

        const results = list
          .find((x) => !x.attributes)
          .children.find((x) => x.name === "div" && !x.attributes).children;

        json.Name = results[0].children[0].children
          .find((x) => x.id?.startsWith("coreui"))
          .children.find((x) => x.name === "h3").children[0];

        json.Banner = results[0].children[0].children
          .find((x) => !x.id?.startsWith("coreui"))
          .children[0].children.find((x) => x.name === "img")
          .attributes["data-src"].replaceAll("amp;", "")
          .replace(/w=162/, "w=600")
          .replace(/h=300/, "h=900");
        return json;
      })()
    );
  }
  return arr;
}

export { getInstalledGames };

File: src-tauri/src/launchers/xbox_games.rs
use std::process::Command;
use serde::{Deserialize, Serialize};
use crate::{launchers::GameObject, operations::custom_fs::read_file, modules::banners};
use reqwest::{Client, Url};
use scraper::{Html, Selector};

#[derive(Serialize, Deserialize)]
pub struct XboxGame {
    #[serde(rename = "Name")]
    name: String,
    #[serde(rename = "PackageFullName")]
    package_full_name: String,
    #[serde(rename = "InstallLocation", default)]
    install_location: String,
    #[serde(rename = "IsFramework")]
    is_framework: String,
    #[serde(rename = "PackageFamilyName")]
    package_family_name: String,
    #[serde(rename = "PublisherId")]
    publisher_id: String,
    #[serde(rename = "SignatureKind")]
    signature_kind: String,
    banner: String,
}

pub async fn get_installed_games() -> Vec<GameObject> {
    let output = Command::new("cmd")
        .args(&[
            "/C",
            "C:/Windows/System32/WindowsPowerShell/v1.0/powershell.exe",
            "get-appxpackage",
        ])
        .output()
        .expect("failed to execute process.");

    let games: Vec<XboxGame> = parse_raw_to_json(
        String::from_utf8_lossy(&output.stdout).to_string()
    )
        .await
        .into_iter()
        .filter(
            |x|
                x.is_framework != "True" &&
                x.signature_kind == "Store" &&
                x.publisher_id != "8wekyb3d8bbwe" &&
                !x.package_full_name.starts_with("MicrosoftWindows")
        )
        .collect::<Vec<XboxGame>>();

    let verified = verify_games(games).await;

    let objs = verified.into_iter().map(|x| parse_game_object(x)).collect::<Vec<_>>();

    return vec![];
}

pub async fn parse_raw_to_json(stdout: String) -> Vec<XboxGame> {
    let mut arr: Vec<XboxGame> = vec![];
    const VALUES: [&str;7] = ["Name", "PackageFullName", "InstallLocation", "IsFramework", "PackageFamilyName", "PublisherId", "SignatureKind"];

    stdout
    .split("\r\n\r\n")
    .collect::<Vec<_>>()[1..]
    .iter()
    .map(
        |x| x
            .split("\r\n")
            .filter(|y| !y.to_string().ends_with(" ") && VALUES.contains(&y.split(":").collect::<Vec<_>>()[0].trim()))
            .map(
                |y| {
                    let splited = y.splitn(3, ":").collect::<Vec<_>>();
                    return splited.iter().enumerate().map(
                        |(i, z)| {
                            let mut val = z.trim().to_owned();
                            if splited.len() == 3 && i != 0 {
                                if i == 2 {
                                    val = format!("{}\"", val);
                                }
                                else if i != 2 {
                                    val = format!("\"{}", val);
                                }
                            }
                            else {
                                val = format!("\"{}\"", val);
                            }
                            val
                        }
                    ).collect::<Vec<_>>().join(":");
                }
            ).collect::<Vec<_>>().join(",\n").replace("\\", "/")
    )
    .filter(|x| x.len() > 20)
    .for_each(|x| {
        arr.push(serde_json::from_str(&("{".to_string() + &x + &",\n\"banner\":\"\"}".to_string())).unwrap());
    });

    return arr;
}

pub async fn verify_games(games: Vec<XboxGame>) -> Vec<XboxGame> {
    let mut arr = vec![];

    for mut game in games {
        game.name = game.name.split(".").collect::<Vec<_>>()[1].to_string();
        let client = Client::new();
        let response = client
        .get(
            Url::parse(&format!("https://www.microsoft.com/en-in/search/shop/games?q={}&devicetype=pc", game.name)).unwrap()
        )
        .send()
        .await
        .expect("");

        let text = &response.text().await.unwrap();
        let document = Html::parse_document(text);
        let selector = Selector::parse(&"div[id=\"shopDetailsWrapper\"]").unwrap();

        let selected = document.select(&selector).collect::<Vec<_>>()[0];
        let ul_element = selected
        .child_elements().collect::<Vec<_>>()[0]
        .child_elements().collect::<Vec<_>>()[2]
        .child_elements().collect::<Vec<_>>();
        // .child_elements().collect::<Vec<_>>();

        if ul_element.len() == 0 {
            continue;
        }

        let li_element;
        if ul_element[0].first_child().is_some() {
            li_element = ul_element[0]
            .child_elements().collect::<Vec<_>>()[0];
        }
        else {
            continue;
        }

        let title = li_element
        .child_elements().collect::<Vec<_>>()[0]
        .child_elements().collect::<Vec<_>>()[0]
        .child_elements().collect::<Vec<_>>()[1]
        .child_elements().collect::<Vec<_>>()[0]
        .child_elements().collect::<Vec<_>>()[0].inner_html();

        let banner = li_element
        .child_elements().collect::<Vec<_>>()[0]
        .child_elements().collect::<Vec<_>>()[0]
        .child_elements().collect::<Vec<_>>()[0]
        .child_elements().collect::<Vec<_>>()[0]
        .child_elements().collect::<Vec<_>>().last()
        .unwrap()
        .attr("src")
        .unwrap();

        game.name = title;
        game.banner = banner.to_string();
        arr.push(game);
    }

    return arr;
}

// Not complete
pub fn parse_game_object(game: XboxGame) -> Option<GameObject> {
    if game.install_location.len() == 0 {
        return None;
    }
    let gameobject: GameObject;
    let regex = Regex::new();
    let location = &game.install_location;
    let manifeststr = read_file(location.to_string() + "/AppxManifest.xml").unwrap();
    let manifest = manifeststr
    .split("\r\n")
    .find(|x| x.trim().starts_with("<Application "))
    .unwrap();
    // .split(/<[/]{0,1}Application[>]{0,1}/);
    let executable = manifest.split("\"")[1].to_string();
    gameobject = GameObject::new(
        banners::get_banner(&game.name, "", "XboxGames", &game.banner),
        executable,
        location.to_string(),
        game.name,
        game.package_full_name,
        "".to_string(),
        0,
        "".to_string(),
        "XboxGames".to_string(),
        vec![],
    );
    return Some(gameobject);
}